analysis.ml

```diff
@@ -48,42 +58,40 @@ let dataflow_analysis (next : pc -> pc list)
         | Some new_state ->
             program_state.(pc) <- merged;
             let updated = update pc new_state in
-            let new_work = List.map (fun pc' -> (updated, pc')) (next pc) in
+            let continue = next.(pc) in
+            let new_work = List.map (fun pc' -> (updated, pc')) continue in
             work (new_work @ rest)
         end
   in
   work init_state;
   program_state
 
-let exits (instrs : instruction_stream) =
-  let rec exits pc : Pc.t list =
-    if Array.length instrs = pc then []
-    else
-      let is_exit = successors instrs pc = [] in
-      if is_exit then pc :: exits (pc + 1) else exits (pc + 1)
-  in
-  exits 0
-
-let forward_analysis_from init_pos init_state seg =
-  let successors pc = successors seg pc in
-  dataflow_analysis successors [(init_state, init_pos)] seg
+exception UnreachableCode of pc
 
-let forward_analysis init_state seg =
-  forward_analysis_from 0 init_state seg
+let make_total result =
+  fun pc ->
+    match result.(pc) with
+    | None -> raise (UnreachableCode pc)
+    | Some res -> res
 
-let backwards_analysis init_state instrs =
-  let exits = exits instrs in
-  let init_state = List.map (fun pc -> (init_state, pc)) exits in
-  let preds = predecessors instrs in
-  let predecessors pc = preds.(pc) in
-  dataflow_analysis predecessors init_state instrs
+let forward_analysis init_state seg merge update =
+  let successors = successors seg in
+  let starts = starts seg in
+  assert (starts != []);
+  let init = List.map (fun pos -> (init_state, pos)) starts in
+  make_total (dataflow_analysis successors init seg merge update)
```

Comment:
One thing I considered while reading this code is to actually make `starts` return the set of instructions without predecessors, dually to `stops`. This would give the guarantee that the analysis is well-defined on all pcs.Then, we could have a `make_total` function that would fail with an exception (or even an assertion failure) at analysis time, rather than at analysis-query time:

```
let make_total result =
    let of_option = function
    | None -> assert false
    | Some res -> res
    in
    let result = Array.map of_option result in
    fun pc -> result.(pc)
```

and further simplify the caller's code.

I am not completely sure this whole suggestion is a good idea. (But decoupling analyses in general from dead-code elimination could be a reasonable choice.)